File: code/java/ch09/practice/Practice01LowerBound.java
package ch09.practice;

import java.util.*;

/**
 * Practice 01: Lower Bound
 * ==============================
 * Chapter 9: Finding Needles — The Power of Searching
 *
 * PROBLEM: Given a SORTED array, find the first index where
 *          arr[i] >= target. Return arr.length if all elements
 *          are smaller than target.
 *
 * ALGORITHM: Binary search variant — narrow the range to find the
 *            insertion point for the target.
 *
 * EXAMPLES:
 *   solve(new int[]{1,3,5,7,9}, 5)  = 2  (arr[2]=5 >= 5)
 *   solve(new int[]{1,3,5,7,9}, 4)  = 2  (arr[2]=5 >= 4)
 *   solve(new int[]{1,3,5,7,9}, 10) = 5  (all smaller)
 *
 * CONSTRAINTS: 0 <= arr.length <= 10^5, array is sorted ascending
 *
 * INSTRUCTIONS: Replace the body of solve() with your solution.
 */
public class Practice01LowerBound {
    public static int solve(int[] arr, int target) {
        // TODO: Replace this with your solution
        return 0;
    }

    // ── Do not change anything below this line ──────────────────────
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String line = sc.nextLine().trim();
        int[] arr;
        if (line.isEmpty()) {
            arr = new int[0];
        } else {
            String[] parts = line.split("\\s+");
            arr = new int[parts.length];
            for (int i = 0; i < parts.length; i++) arr[i] = Integer.parseInt(parts[i]);
        }
        int target = Integer.parseInt(sc.nextLine().trim());
        System.out.println(solve(arr, target));
        sc.close();
    }
}

File: code/python/ch09/practice/practice_01_lower_bound.py
"""
Practice 1: Lower Bound
==============================
Chapter 9: Finding Needles — The Power of Searching

PROBLEM
-------
Given a sorted array of integers and a target value, find the lower
bound: the index of the first element that is greater than or equal
to the target. If all elements are smaller than the target, return
len(arr) (one past the last index).

INPUT FORMAT
------------
First line: space-separated integers (a sorted array).
Second line: a single integer (the target).

OUTPUT FORMAT
-------------
A single integer: the first index where arr[i] >= target.

CONSTRAINTS
-----------
- 0 <= len(arr) <= 10^5
- -10^9 <= arr[i] <= 10^9
- The array is sorted in non-decreasing order.

EXAMPLES
--------
Input:
  1 3 5 7 9
  5
Output: 2

Input:
  1 3 5 7 9
  4
Output: 2

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
The main block at the bottom handles input/output — don't change it.
"""


def solve(arr: list[int], target: int) -> int:
    """Return the first index where arr[i] >= target."""
    pass  # TODO: Replace this with your solution


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    data = list(map(int, input().split()))
    target = int(input())
    print(solve(data, target))
